File: shared/delivery.mjs
// Cột "Phân phối" giống Ads Manager: không chỉ là bật/tắt của chính mục đó mà còn xét các nhóm QC bên trong
// (vd camp đang bật nhưng mọi nhóm QC đều tắt → "Nhóm quảng cáo đang tắt").
// rank: thứ tự khi sắp xếp theo cột Phân phối (đang chạy lên đầu).
export const DELIVERY = {
  active: { label: 'Đang hoạt động', tone: 'success', rank: 0, running: true },
  learning: { label: 'Đang học', tone: 'info', rank: 1, running: true },
  review: { label: 'Đang xét duyệt', tone: 'warning', rank: 2 },
  issues: { label: 'Có vấn đề', tone: 'warning', rank: 3 },
  scheduled: { label: 'Đã lên lịch', tone: 'info', rank: 4 },
  adsetsOff: { label: 'Nhóm quảng cáo đang tắt', tone: 'neutral', rank: 5 },
  campaignOff: { label: 'Chiến dịch đang tắt', tone: 'neutral', rank: 6 },
  rejected: { label: 'Bị từ chối', tone: 'danger', rank: 7 },
  completed: { label: 'Hoàn tất', tone: 'neutral', rank: 8 },
  off: { label: 'Tắt', tone: 'neutral', rank: 9 },
  archived: { label: 'Lưu trữ', tone: 'neutral', rank: 10 },
  deleted: { label: 'Đã xoá', tone: 'neutral', rank: 11 },
}

// Phần chung cho camp và nhóm QC: trạng thái do chính mục đó quyết định. Trả về '' nếu cần xét tiếp.
function own(o) {
  const e = o.effective
  if (e === 'DELETED') return 'deleted'
  if (e === 'ARCHIVED') return 'archived'
  if (o.status === 'PAUSED') return 'off'
  if (e === 'CAMPAIGN_PAUSED') return 'campaignOff'
  if (e === 'ADSET_PAUSED') return 'adsetsOff'
  if (e === 'DISAPPROVED') return 'rejected'
  if (e === 'PENDING_REVIEW' || e === 'IN_PROCESS') return 'review'
  if (e === 'WITH_ISSUES') return 'issues'
  return ''
}

function adsetDelivery(a, now) {
  const k = own(a)
  if (k) return k
  if (a.endTime && a.endTime < now) return 'completed'
  if (a.startTime && a.startTime > now) return 'scheduled'
  if (a.learning) return 'learning'
  return a.effective === 'ACTIVE' ? 'active' : 'off'
}

function campaignDelivery(c, adsets, now) {
  const k = own(c)
  if (k) return k
  if (!adsets.length) return c.effective === 'ACTIVE' ? 'active' : 'off' // chưa có dữ liệu nhóm QC (vd dữ liệu giả)
  const ks = adsets.map((a) => adsetDelivery(a, now))
  const has = (x) => ks.includes(x)
  if (has('active') || has('learning')) return 'active'
  for (const x of ['review', 'issues', 'scheduled']) if (has(x)) return x
  const gone = (x) => ['completed', 'archived', 'deleted'].includes(x)
  if (has('completed') && ks.every(gone)) return 'completed'
  if (has('rejected') && ks.every((x) => x === 'rejected' || gone(x))) return 'rejected'
  return 'adsetsOff'
}

// Map id → khoá trong DELIVERY cho toàn bộ danh sách (camp + nhóm QC)
export function deliveryMap(objs, now = Date.now()) {
  const byCamp = {}
  for (const o of objs) if (o.level === 'adset') (byCamp[o.campaignId] ||= []).push(o)
  const out = {}
  for (const o of objs) out[o.id] = o.level === 'adset' ? adsetDelivery(o, now) : campaignDelivery(o, byCamp[o.id] || [], now)
  return out
}

File: backend-java/src/main/java/com/fbads/facebook/Delivery.java
package com.fbads.facebook;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Cột "Phân phối" giống Ads Manager (bản Java của shared/delivery.mjs): không chỉ bật/tắt của chính mục đó mà còn xét
 * các nhóm QC bên trong (camp bật nhưng mọi nhóm QC tắt → "Nhóm quảng cáo đang tắt", KHÔNG tính là đang chạy).
 */
public final class Delivery {
    public record Info(String label, boolean running) {}

    public static final Map<String, Info> INFO = Map.ofEntries(
            Map.entry("active", new Info("Đang hoạt động", true)),
            Map.entry("learning", new Info("Đang học", true)),
            Map.entry("review", new Info("Đang xét duyệt", false)),
            Map.entry("issues", new Info("Có vấn đề", false)),
            Map.entry("scheduled", new Info("Đã lên lịch", false)),
            Map.entry("adsetsOff", new Info("Nhóm quảng cáo đang tắt", false)),
            Map.entry("campaignOff", new Info("Chiến dịch đang tắt", false)),
            Map.entry("rejected", new Info("Bị từ chối", false)),
            Map.entry("completed", new Info("Hoàn tất", false)),
            Map.entry("off", new Info("Tắt", false)),
            Map.entry("archived", new Info("Lưu trữ", false)),
            Map.entry("deleted", new Info("Đã xoá", false)));

    private Delivery() {}

    private static String own(AdObject o) {
        String e = o.effective == null ? "" : o.effective;
        if (e.equals("DELETED")) return "deleted";
        if (e.equals("ARCHIVED")) return "archived";
        if ("PAUSED".equals(o.status)) return "off";
        if (e.equals("CAMPAIGN_PAUSED")) return "campaignOff";
        if (e.equals("ADSET_PAUSED")) return "adsetsOff";
        if (e.equals("DISAPPROVED")) return "rejected";
        if (e.equals("PENDING_REVIEW") || e.equals("IN_PROCESS")) return "review";
        if (e.equals("WITH_ISSUES")) return "issues";
        return "";
    }

    private static String adset(AdObject a, long now) {
        String k = own(a);
        if (!k.isEmpty()) return k;
        if (a.endTime != null && a.endTime != 0 && a.endTime < now) return "completed";
        if (a.startTime != null && a.startTime != 0 && a.startTime > now) return "scheduled";
        if (a.learning) return "learning";
        return a.isActive() ? "active" : "off";
    }

    private static String campaign(AdObject c, List<AdObject> adsets, long now) {
        String k = own(c);
        if (!k.isEmpty()) return k;
        if (adsets.isEmpty()) return c.isActive() ? "active" : "off"; // chưa có dữ liệu nhóm QC (vd dữ liệu giả)
        List<String> ks = adsets.stream().map(a -> adset(a, now)).toList();
        if (ks.contains("active") || ks.contains("learning")) return "active";
        for (String x : List.of("review", "issues", "scheduled")) if (ks.contains(x)) return x;
        java.util.function.Predicate<String> gone = x -> x.equals("completed") || x.equals("archived") || x.equals("deleted");
        if (ks.contains("completed") && ks.stream().allMatch(gone)) return "completed";
        if (ks.contains("rejected") && ks.stream().allMatch(x -> x.equals("rejected") || gone.test(x))) return "rejected";
        return "adsetsOff";
    }

    /** id → khoá trạng thái phân phối cho toàn bộ danh sách */
    public static Map<String, String> map(List<AdObject> objs, long now) {
        Map<String, List<AdObject>> byCamp = new HashMap<>();
        for (AdObject o : objs) if ("adset".equals(o.level)) byCamp.computeIfAbsent(o.campaignId, k -> new ArrayList<>()).add(o);
        Map<String, String> out = new HashMap<>();
        for (AdObject o : objs) out.put(o.id, "adset".equals(o.level) ? adset(o, now) : campaign(o, byCamp.getOrDefault(o.id, List.of()), now));
        return out;
    }

    /** Hỏi nhanh "mục này có đang thật sự chạy không" + nhãn trạng thái (dùng cho rule) */
    public static final class View {
        private final Map<String, String> map;

        public View(List<AdObject> objs) { this.map = Delivery.map(objs, System.currentTimeMillis()); }

        public boolean running(AdObject o) { Info i = INFO.get(map.get(o.id)); return i != null && i.running(); }

        public String label(AdObject o) { Info i = INFO.get(map.get(o.id)); return i == null ? "" : i.label(); }
    }
}
